File: sdkwork-rtc-sdk-java/src/main/java/com/sdkwork/rtc/metadata/RtcCapabilityCatalog.java
package com.sdkwork.rtc.metadata;

import java.util.List;
import java.util.Optional;

public final class RtcCapabilityCatalog {

  public static final List<Entry> ENTRIES = List.of(
      new Entry("session", "required-baseline", "cross-surface"),
      new Entry("join", "required-baseline", "runtime-bridge"),
      new Entry("publish", "required-baseline", "runtime-bridge"),
      new Entry("subscribe", "required-baseline", "runtime-bridge"),
      new Entry("mute", "required-baseline", "runtime-bridge"),
      new Entry("basic-events", "required-baseline", "runtime-bridge"),
      new Entry("health", "required-baseline", "control-plane"),
      new Entry("unwrap", "required-baseline", "runtime-bridge"),
      new Entry("screen-share", "optional-advanced", "runtime-bridge"),
      new Entry("recording", "optional-advanced", "control-plane"),
      new Entry("cloud-mix", "optional-advanced", "control-plane"),
      new Entry("cdn-relay", "optional-advanced", "control-plane"),
      new Entry("data-channel", "optional-advanced", "runtime-bridge"),
      new Entry("transcription", "optional-advanced", "control-plane"),
      new Entry("beauty", "optional-advanced", "runtime-bridge"),
      new Entry("spatial-audio", "optional-advanced", "runtime-bridge"),
      new Entry("e2ee", "optional-advanced", "runtime-bridge")
  );

public static List<Entry> getRtcCapabilityCatalog() {
    return ENTRIES;
  }

  public static Optional<Entry> getRtcCapabilityDescriptor(String capabilityKey) {
    for (var entry : ENTRIES) {
      if (entry.capabilityKey().equals(capabilityKey)) {
        return Optional.of(entry);
      }
    }

    return Optional.empty();
  }


  private RtcCapabilityCatalog() {
  }

  public record Entry(String capabilityKey, String category, String surface) {
  }
}

File: sdkwork-rtc-sdk-kotlin/src/main/kotlin/com/sdkwork/rtc/metadata/RtcCapabilityCatalog.kt
package com.sdkwork.rtc.metadata

data class RtcCapabilityCatalogEntry(
    val capabilityKey: String,
    val category: String,
    val surface: String,
)

object RtcCapabilityCatalog {
    val entries: List<RtcCapabilityCatalogEntry> = listOf(
        RtcCapabilityCatalogEntry("session", "required-baseline", "cross-surface"),
        RtcCapabilityCatalogEntry("join", "required-baseline", "runtime-bridge"),
        RtcCapabilityCatalogEntry("publish", "required-baseline", "runtime-bridge"),
        RtcCapabilityCatalogEntry("subscribe", "required-baseline", "runtime-bridge"),
        RtcCapabilityCatalogEntry("mute", "required-baseline", "runtime-bridge"),
        RtcCapabilityCatalogEntry("basic-events", "required-baseline", "runtime-bridge"),
        RtcCapabilityCatalogEntry("health", "required-baseline", "control-plane"),
        RtcCapabilityCatalogEntry("unwrap", "required-baseline", "runtime-bridge"),
        RtcCapabilityCatalogEntry("screen-share", "optional-advanced", "runtime-bridge"),
        RtcCapabilityCatalogEntry("recording", "optional-advanced", "control-plane"),
        RtcCapabilityCatalogEntry("cloud-mix", "optional-advanced", "control-plane"),
        RtcCapabilityCatalogEntry("cdn-relay", "optional-advanced", "control-plane"),
        RtcCapabilityCatalogEntry("data-channel", "optional-advanced", "runtime-bridge"),
        RtcCapabilityCatalogEntry("transcription", "optional-advanced", "control-plane"),
        RtcCapabilityCatalogEntry("beauty", "optional-advanced", "runtime-bridge"),
        RtcCapabilityCatalogEntry("spatial-audio", "optional-advanced", "runtime-bridge"),
        RtcCapabilityCatalogEntry("e2ee", "optional-advanced", "runtime-bridge"),
    )

fun getRtcCapabilityCatalog(): List<RtcCapabilityCatalogEntry> = entries

    fun getRtcCapabilityDescriptor(capabilityKey: String): RtcCapabilityCatalogEntry? =
        entries.firstOrNull { it.capabilityKey == capabilityKey }

}
